File: lib/pipeline/dither.ts
import { rgbToOklab } from "../color/color";
import { DEFAULT_DITHER_TEXTURE, handDrawnThresholds, handDrawnThresholdWindow, type DitherTexture } from "./dither-hand-drawn";
import { DITHER_MATRICES } from "./dither-matrices";
import type { RGB } from "../types";

/**
 * Dithering a chart to its palette (G-052): instead of rounding every stitch to its nearest thread, neighbouring
 * stitches take the two threads either side of the colour, so their mixture reads as the shade in between. It is the
 * cross-stitch technique of the same name, and it buys gradients a small palette cannot hold — at the cost of
 * confetti, which the Owner accepted (2026-09-21). See `docs/reviews/2026-09-21-dithering-research.md`.
 *
 * Two families, as the research found them:
 * - **Threshold matrix** (every mode but the last): each stitch compares its position between the two nearest threads
 *   against a repeating matrix. The matrix is the pattern — Bayer, a clustered dot screen, line screens, blue noise —
 *   and adding another is data, not code (D198).
 * - **Error diffusion** (`floyd-steinberg`): round each stitch, then push the colour error onto the stitches not yet
 *   decided, so it averages out over an area. Serpentine, which is what stops the worm artifacts the plain
 *   left-to-right scan makes.
 *
 * Everything works in OKLab, the space the rest of the pipeline compares colours in.
 */

/** The line screens: one pattern with a direction, stored as four ids so files written before G-059 still open. */
export const LINE_DITHER_MODES = ["lines-horizontal", "lines-vertical", "lines-diagonal", "lines-anti-diagonal"] as const;
export type LineDitherMode = (typeof LINE_DITHER_MODES)[number];
export const ORDERED_DITHER_MODES = ["bayer-4", "bayer-8", "clustered-8", "ring-8", ...LINE_DITHER_MODES, "blue-noise-16"] as const;
export type OrderedDitherMode = (typeof ORDERED_DITHER_MODES)[number];
export const DIFFUSION_DITHER_MODES = ["floyd-steinberg", "atkinson"] as const;
/** Marks placed across the chart rather than a tile repeated or an error carried: the third family (G-054). */
export const DRAWN_DITHER_MODES = ["hand-drawn"] as const;
export const DITHER_MODES = ["off", ...ORDERED_DITHER_MODES, ...DIFFUSION_DITHER_MODES, ...DRAWN_DITHER_MODES] as const;
export type DiffusionDitherMode = (typeof DIFFUSION_DITHER_MODES)[number];
export type DitherMode = (typeof DITHER_MODES)[number];

export function isDithered(mode: DitherMode | undefined): mode is Exclude<DitherMode, "off"> {
  return mode !== undefined && mode !== "off";
}

/** Whether a pattern is one of the line screens, which the pane offers as a single option with a direction. */
export function isLinesMode(mode: DitherMode | undefined): mode is LineDitherMode {
  return mode !== undefined && (LINE_DITHER_MODES as readonly string[]).includes(mode);
}

/** Whether a pattern draws marks across the whole chart instead of repeating a tile or carrying an error. */
export function isDrawnMode(mode: DitherMode | undefined): mode is (typeof DRAWN_DITHER_MODES)[number] {
  return mode !== undefined && (DRAWN_DITHER_MODES as readonly string[]).includes(mode);
}

/** Whether a pattern carries the error to the stitches after it, rather than reading a threshold matrix. */
export function isDiffusionMode(mode: Exclude<DitherMode, "off">): mode is DiffusionDitherMode {
  return (DIFFUSION_DITHER_MODES as readonly string[]).includes(mode);
}

/** The two nearest palette entries to a colour, nearest first. With one entry both are it. */
function twoNearest(paletteOklab: Float64Array, count: number, l: number, a: number, b: number): [number, number] {
  let best = 0;
  let bestDist = Infinity;
  let second = 0;
  let secondDist = Infinity;
  for (let c = 0; c < count; c++) {
    const o = c * 3;
    const dl = l - paletteOklab[o];
    const da = a - paletteOklab[o + 1];
    const db = b - paletteOklab[o + 2];
    const d = dl * dl + da * da + db * db;
    if (d < bestDist) {
      secondDist = bestDist;
      second = best;
      bestDist = d;
      best = c;
    } else if (d < secondDist) {
      secondDist = d;
      second = c;
    }
  }
  return [best, secondDist === Infinity ? best : second];
}

/** Interleaved OKLab of a palette. */
function paletteToOklab(palette: readonly RGB[]): Float64Array {
  const out = new Float64Array(palette.length * 3);
  palette.forEach((rgb, i) => {
    const [l, a, b] = rgbToOklab(rgb);
    out[i * 3] = l;
    out[i * 3 + 1] = a;
    out[i * 3 + 2] = b;
  });
  return out;
}

/**
 * Where a colour sits between two palette entries: 0 at the first, 1 at the second, clamped. This is what the
 * threshold decides between, and it is why an unevenly spaced palette — a thread line — dithers correctly here: the
 * mixture is measured between the two threads that actually bracket the colour, not against an assumed even ramp.
 */
function positionBetween(paletteOklab: Float64Array, first: number, second: number, l: number, a: number, b: number): number {
  const fo = first * 3;
  const so = second * 3;
  const vl = paletteOklab[so] - paletteOklab[fo];
  const va = paletteOklab[so + 1] - paletteOklab[fo + 1];
  const vb = paletteOklab[so + 2] - paletteOklab[fo + 2];
  const lengthSquared = vl * vl + va * va + vb * vb;
  if (lengthSquared === 0) return 0;
  const t = ((l - paletteOklab[fo]) * vl + (a - paletteOklab[fo + 1]) * va + (b - paletteOklab[fo + 2]) * vb) / lengthSquared;
  return t <= 0 ? 0 : t >= 1 ? 1 : t;
}

/** One label per cell, from the matrix named by `mode`. */
function orderedDither(cellOklab: Float64Array, width: number, height: number, palette: readonly RGB[], mode: OrderedDitherMode): Uint8Array {
  const matrix = DITHER_MATRICES[mode];
  if (!matrix) throw new Error(`Unknown dither pattern "${mode}".`);
  const size = matrix.length;
  const scale = size * size;
  const paletteOklab = paletteToOklab(palette);
  const labels = new Uint8Array(width * height);

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = y * width + x;
      const o = i * 3;
      const l = cellOklab[o];
      const a = cellOklab[o + 1];
      const b = cellOklab[o + 2];
      const [first, second] = twoNearest(paletteOklab, palette.length, l, a, b);
      const t = positionBetween(paletteOklab, first, second, l, a, b);
      // The rank's midpoint, so a tone of exactly 0.5 splits the matrix evenly instead of favouring one side.
      const threshold = (matrix[y % size][x % size] + 0.5) / scale;
      labels[i] = t > threshold ? second : first;
    }
  }
  return labels;
}

/**
 * One label per cell from a threshold field covering the whole chart, rather than a tile repeated across it. The
 * decision is the ordered one — the field only says where each cell sits in its mark.
 */
function drawnDither(cellOklab: Float64Array, width: number, height: number, palette: readonly RGB[], texture: DitherTexture): Uint8Array {
  const thresholds = handDrawnThresholds(width, height, texture);
  const paletteOklab = paletteToOklab(palette);
  const labels = new Uint8Array(width * height);

  for (let i = 0; i < width * height; i++) {
    const o = i * 3;
    const l = cellOklab[o];
    const a = cellOklab[o + 1];
    const b = cellOklab[o + 2];
    const [first, second] = twoNearest(paletteOklab, palette.length, l, a, b);
    const t = positionBetween(paletteOklab, first, second, l, a, b);
    labels[i] = t > thresholds[i] ? second : first;
  }
  return labels;
}

/**
 * The error-diffusion kernels, as `[dx, dy, weight]` taps ahead of the cell being decided. `dx` is mirrored on a
 * right-to-left row, so a serpentine scan spreads the error the same way in both directions.
 *
 * Floyd–Steinberg passes all of the error on. Atkinson passes only six eighths and drops the rest, which is what
 * gives it its look: the error never accumulates enough to break a near-black or near-white area, so the extremes
 * stay flat, and the stitches it does place clump (D200).
 */
const DIFFUSION_KERNELS: Record<DiffusionDitherMode, readonly (readonly [number, number, number])[]> = {
  "floyd-steinberg": [
    [1, 0, 7 / 16],
    [-1, 1, 3 / 16],
    [0, 1, 5 / 16],
    [1, 1, 1 / 16],
  ],
  atkinson: [
    [1, 0, 1 / 8],
    [2, 0, 1 / 8],
    [-1, 1, 1 / 8],
    [0, 1, 1 / 8],
    [1, 1, 1 / 8],
    [0, 2, 1 / 8],
  ],
};

/** One label per cell, each rounded to its nearest thread with the error carried to the stitches not yet decided. */
function errorDiffusionDither(cellOklab: Float64Array, width: number, height: number, palette: readonly RGB[], mode: DiffusionDitherMode): Uint8Array {
  const paletteOklab = paletteToOklab(palette);
  const labels = new Uint8Array(width * height);
  // A working copy, because a cell's colour is its own plus whatever error reached it.
  const working = Float64Array.from(cellOklab);

  for (let y = 0; y < height; y++) {
    // Serpentine: every other row runs right to left, which breaks up the worms a one-way scan leaves behind.
    const leftToRight = y % 2 === 0;
    for (let step = 0; step < width; step++) {
      const x = leftToRight ? step : width - 1 - step;
      const i = y * width + x;
      const o = i * 3;
      const [best] = twoNearest(paletteOklab, palette.length, working[o], working[o + 1], working[o + 2]);
      labels[i] = best;
      const po = best * 3;
      const el = working[o] - paletteOklab[po];
      const ea = working[o + 1] - paletteOklab[po + 1];
      const eb = working[o + 2] - paletteOklab[po + 2];

      const spread = (nx: number, ny: number, weight: number) => {
        if (nx < 0 || nx >= width || ny < 0 || ny >= height) return;
        const no = (ny * width + nx) * 3;
        working[no] += el * weight;
        working[no + 1] += ea * weight;
        working[no + 2] += eb * weight;
      };
      const ahead = leftToRight ? 1 : -1;
      for (const [dx, dy, weight] of DIFFUSION_KERNELS[mode]) spread(x + ahead * dx, y + dy, weight);
    }
  }
  return labels;
}

/** The tone the swatch's ramp shows at a row: dark at the top, light at the bottom, never quite either. */
export function drawnRampTone(row: number, height: number): number {
  return height <= 1 ? 0.5 : 0.15 + (0.7 * row) / (height - 1);
}

/**
 * What a chart of `chartWidth` × `chartHeight` would show in its top-left window, for a dark-to-light ramp between
 * two threads: the editor's swatch (G-057).
 *
 * It exists here, beside `ditherToPalette`, because the swatch has to make its stitches the way a chart does — the
 * same threshold field, and the same rule about which of the two nearest threads is the one being placed. A swatch
 * that compared a tone with a threshold directly looked right in the dark half and came out inverted in the light
 * half, where the nearer thread is the light one.
 */
export function drawnRampWindow(
  chartWidth: number,
  chartHeight: number,
  windowWidth: number,
  windowHeight: number,
  palette: readonly RGB[],
  texture: DitherTexture = DEFAULT_DITHER_TEXTURE
): { width: number; height: number; labels: Uint8Array } {
  const { width, height, thresholds } = handDrawnThresholdWindow(chartWidth, chartHeight, windowWidth, windowHeight, texture);
  const labels = new Uint8Array(width * height);
  if (palette.length < 2) return { width, height, labels };
  const paletteOklab = paletteToOklab(palette);
  const from = rgbToOklab(palette[0]);
  const to = rgbToOklab(palette[1]);

  for (let y = 0; y < height; y++) {
    const tone = drawnRampTone(y, height);
    const l = from[0] + tone * (to[0] - from[0]);
    const a = from[1] + tone * (to[1] - from[1]);
    const b = from[2] + tone * (to[2] - from[2]);
    const [first, second] = twoNearest(paletteOklab, palette.length, l, a, b);
    const t = positionBetween(paletteOklab, first, second, l, a, b);
    for (let x = 0; x < width; x++) labels[y * width + x] = t > thresholds[y * width + x] ? second : first;
  }
  return { width, height, labels };
}

/** Every cell's palette entry, dithered by `mode`. The palette is the quantizer's, unchanged. */
export function ditherToPalette(
  cellOklab: Float64Array,
  width: number,
  height: number,
  palette: readonly RGB[],
  mode: Exclude<DitherMode, "off">,
  texture: DitherTexture = DEFAULT_DITHER_TEXTURE
): Uint8Array {
  if (palette.length === 0) return new Uint8Array(width * height);
  if (isDiffusionMode(mode)) return errorDiffusionDither(cellOklab, width, height, palette, mode);
  if (isDrawnMode(mode)) return drawnDither(cellOklab, width, height, palette, texture);
  return orderedDither(cellOklab, width, height, palette, mode);
}

File: rust/cs-core/src/dither.rs
//! Port of `lib/pipeline/dither.ts` (G-052): a chart dithered to its palette, so neighbouring stitches take the two
//! threads either side of a colour instead of every stitch rounding to one. Threshold matrices are the patterns and
//! come from the same generated data the TypeScript reads (D198); `floyd-steinberg` is the error-diffusion family,
//! serpentine, which is what keeps the worm artifacts away.

use crate::color::{rgb_to_oklab, Oklab, Rgb};
use std::collections::HashMap;
use std::sync::OnceLock;

/// Which pattern, or none. The values are `DitherMode`'s in TypeScript.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum DitherMode {
    Off,
    Bayer4,
    Bayer8,
    Clustered8,
    Ring8,
    LinesHorizontal,
    LinesDiagonal,
    BlueNoise16,
    FloydSteinberg,
    Atkinson,
}

impl DitherMode {
    pub fn id(self) -> &'static str {
        match self {
            DitherMode::Off => "off",
            DitherMode::Bayer4 => "bayer-4",
            DitherMode::Bayer8 => "bayer-8",
            DitherMode::Clustered8 => "clustered-8",
            DitherMode::Ring8 => "ring-8",
            DitherMode::LinesHorizontal => "lines-horizontal",
            DitherMode::LinesDiagonal => "lines-diagonal",
            DitherMode::BlueNoise16 => "blue-noise-16",
            DitherMode::FloydSteinberg => "floyd-steinberg",
            DitherMode::Atkinson => "atkinson",
        }
    }

    pub fn is_dithered(self) -> bool {
        self != DitherMode::Off
    }
}

struct Matrix {
    size: usize,
    ranks: Vec<u32>,
}

/// The generated patterns, parsed once: `name<TAB>size<TAB>comma-separated ranks in row-major order`.
fn matrices() -> &'static HashMap<String, Matrix> {
    static M: OnceLock<HashMap<String, Matrix>> = OnceLock::new();
    M.get_or_init(|| {
        let mut out = HashMap::new();
        for line in include_str!("../data/dither-matrices.tsv").lines() {
            let mut parts = line.split('\t');
            let (Some(name), Some(size), Some(ranks)) = (parts.next(), parts.next(), parts.next())
            else {
                continue;
            };
            out.insert(
                name.to_string(),
                Matrix {
                    size: size.parse().expect("matrix size"),
                    ranks: ranks
                        .split(',')
                        .map(|r| r.parse().expect("matrix rank"))
                        .collect(),
                },
            );
        }
        out
    })
}

/// `twoNearest`: the two nearest palette entries, nearest first, with the TypeScript's tie handling.
fn two_nearest(palette: &[f64], count: usize, l: f64, a: f64, b: f64) -> (usize, usize) {
    let mut best = 0usize;
    let mut best_dist = f64::INFINITY;
    let mut second = 0usize;
    let mut second_dist = f64::INFINITY;
    for c in 0..count {
        let o = c * 3;
        let dl = l - palette[o];
        let da = a - palette[o + 1];
        let db = b - palette[o + 2];
        let d = dl * dl + da * da + db * db;
        if d < best_dist {
            second_dist = best_dist;
            second = best;
            best_dist = d;
            best = c;
        } else if d < second_dist {
            second_dist = d;
            second = c;
        }
    }
    (
        best,
        if second_dist.is_infinite() {
            best
        } else {
            second
        },
    )
}

fn palette_to_oklab(palette: &[Rgb]) -> Vec<f64> {
    let mut out = vec![0f64; palette.len() * 3];
    for (i, &rgb) in palette.iter().enumerate() {
        let lab: Oklab = rgb_to_oklab(rgb);
        out[i * 3..i * 3 + 3].copy_from_slice(&lab);
    }
    out
}

/// `positionBetween`: where a colour sits between two palette entries, 0 at the first and 1 at the second.
fn position_between(palette: &[f64], first: usize, second: usize, l: f64, a: f64, b: f64) -> f64 {
    let (fo, so) = (first * 3, second * 3);
    let vl = palette[so] - palette[fo];
    let va = palette[so + 1] - palette[fo + 1];
    let vb = palette[so + 2] - palette[fo + 2];
    let length_squared = vl * vl + va * va + vb * vb;
    if length_squared == 0.0 {
        return 0.0;
    }
    let t = ((l - palette[fo]) * vl + (a - palette[fo + 1]) * va + (b - palette[fo + 2]) * vb)
        / length_squared;
    if t <= 0.0 {
        0.0
    } else if t >= 1.0 {
        1.0
    } else {
        t
    }
}

fn ordered(
    cell_oklab: &[f64],
    width: usize,
    height: usize,
    palette: &[Rgb],
    mode: DitherMode,
) -> Vec<u8> {
    let matrix = matrices()
        .get(mode.id())
        .unwrap_or_else(|| panic!("unknown dither pattern {}", mode.id()));
    let size = matrix.size;
    let scale = (size * size) as f64;
    let palette_oklab = palette_to_oklab(palette);
    let mut labels = vec![0u8; width * height];

    for y in 0..height {
        for x in 0..width {
            let i = y * width + x;
            let o = i * 3;
            let (l, a, b) = (cell_oklab[o], cell_oklab[o + 1], cell_oklab[o + 2]);
            let (first, second) = two_nearest(&palette_oklab, palette.len(), l, a, b);
            let t = position_between(&palette_oklab, first, second, l, a, b);
            let threshold = (matrix.ranks[(y % size) * size + (x % size)] as f64 + 0.5) / scale;
            labels[i] = if t > threshold {
                second as u8
            } else {
                first as u8
            };
        }
    }
    labels
}

/// The error-diffusion kernels as `(dx, dy, weight)` taps, mirrored in `dx` on a right-to-left row. Floyd-Steinberg
/// passes all of the error on; Atkinson passes six eighths and drops the rest, which keeps near-black and near-white
/// areas flat and makes the stitches it does place clump (D200). Mirrors `DIFFUSION_KERNELS` in `dither.ts`.
const FLOYD_STEINBERG: [(i64, i64, f64); 4] = [
    (1, 0, 7.0 / 16.0),
    (-1, 1, 3.0 / 16.0),
    (0, 1, 5.0 / 16.0),
    (1, 1, 1.0 / 16.0),
];
const ATKINSON: [(i64, i64, f64); 6] = [
    (1, 0, 1.0 / 8.0),
    (2, 0, 1.0 / 8.0),
    (-1, 1, 1.0 / 8.0),
    (0, 1, 1.0 / 8.0),
    (1, 1, 1.0 / 8.0),
    (0, 2, 1.0 / 8.0),
];

fn error_diffusion(
    cell_oklab: &[f64],
    width: usize,
    height: usize,
    palette: &[Rgb],
    mode: DitherMode,
) -> Vec<u8> {
    let kernel: &[(i64, i64, f64)] = match mode {
        DitherMode::Atkinson => &ATKINSON,
        _ => &FLOYD_STEINBERG,
    };
    let palette_oklab = palette_to_oklab(palette);
    let mut labels = vec![0u8; width * height];
    let mut working = cell_oklab.to_vec();

    for y in 0..height {
        let left_to_right = y % 2 == 0;
        for step in 0..width {
            let x = if left_to_right {
                step
            } else {
                width - 1 - step
            };
            let i = y * width + x;
            let o = i * 3;
            let (best, _) = two_nearest(
                &palette_oklab,
                palette.len(),
                working[o],
                working[o + 1],
                working[o + 2],
            );
            labels[i] = best as u8;
            let po = best * 3;
            let el = working[o] - palette_oklab[po];
            let ea = working[o + 1] - palette_oklab[po + 1];
            let eb = working[o + 2] - palette_oklab[po + 2];

            let ahead: i64 = if left_to_right { 1 } else { -1 };
            let mut spread = |nx: i64, ny: i64, weight: f64| {
                if nx < 0 || nx >= width as i64 || ny < 0 || ny >= height as i64 {
                    return;
                }
                let no = (ny as usize * width + nx as usize) * 3;
                working[no] += el * weight;
                working[no + 1] += ea * weight;
                working[no + 2] += eb * weight;
            };
            let (xi, yi) = (x as i64, y as i64);
            for &(dx, dy, weight) in kernel {
                spread(xi + ahead * dx, yi + dy, weight);
            }
        }
    }
    labels
}

/// `ditherToPalette`: every cell's palette entry, dithered by `mode`.
pub fn dither_to_palette(
    cell_oklab: &[f64],
    width: usize,
    height: usize,
    palette: &[Rgb],
    mode: DitherMode,
) -> Vec<u8> {
    if palette.is_empty() {
        return vec![0u8; width * height];
    }
    match mode {
        DitherMode::Off => panic!("dither_to_palette called with Off"),
        DitherMode::FloydSteinberg | DitherMode::Atkinson => {
            error_diffusion(cell_oklab, width, height, palette, mode)
        }
        _ => ordered(cell_oklab, width, height, palette, mode),
    }
}
